src/madpack/create_changelist.py

```diff
@@ -0,0 +1,239 @@
+        print "Git diff failed. Please ensure that branches/tags are fetched."
+        raise SystemExit
+
+    f = open("/tmp/madlib_tmp_cl.yaml", "w")
+    f.write(
+"""# ------------------------------------------------------------------------------
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# ------------------------------------------------------------------------------
+""")
+
+    f.write(
+    """
+# Changelist for MADlib version {old_vers} to {new_vers}
+
+# This file contains all changes that were introduced in a new version of
+# MADlib. This changelist is used by the upgrade script to detect what objects
+# should be upgraded (while retaining all other objects from the previous version)
+
+# New modules (actually .sql_in files) added in upgrade version
+# For these files the sql_in code is retained as is with the functions in the
+# file installed on the upgrade version. All other files (that don't have
+# updates), are cleaned up to remove object replacements
+""".format(**locals()))
+
+    # Find the new .sql_in files that are not in test folders
+    f.write("new module:\n")
+    with open('/tmp/madlib_tmp_nm.txt') as fp:
+        for line in fp:
+            if 'sql_in' in line and '/test/' not in line:
+                 f.write('    ' + line.split('/')[5].split('.')[0]+':\n')
+
+    # Find the changed types and keep a list for future use
+    os.system("psql {0} -f diff_udt.sql > /tmp/madlib_tmp_udt.txt".format(database))
+
+    f.write("\n# Changes in the types (UDT) including removal and modification\n")
+    f.write("udt:\n")
+    udt_list=[]
+    with open('/tmp/madlib_tmp_udt.txt') as fp:
+        for line in fp:
+           if 'UDT' in line and '[]' not in line:
+                ch_type = line.split('|')[1].strip()
+                udt_list.append(ch_type)
+                f.write('    ' + ch_type +":\n")
+
+    # Find the list of UDFs and UDAs
+    # There are two main sources for these lists.
+    # 1. The functions that actually got changed
+    # 2. The functions that depend on a changed type
+
+    # We will keep two lists (for UDF and UDA) and fill them as we parse the
+    # output of diff functions
+
+    udf_list=[]
+    uda_list=[]
+    current_list = udf_list
```

Comment:
Do we really need the current_list variable? The code will look a bit simpler if we just append to the right list whenever necessary.

Reply:
We don't "need" the current list but it helps us to reduce the number of copy/pasted lines.